relay-event-normalization/src/sdk_version.rs

```
@@ -0,0 +1,148 @@
+use std::num::ParseIntError;
+
+const DEFAULT_VERSION_IF_EMPTY: usize = 0;
+
+/// Represents an SDK Version using the semvar versioning, meaning MAJOR.MINOR.PATCH.
+/// An optional release type can be specified, then it becomes
+/// MAJOR.MINOR.PATCH-(alpha|beta).RELEASE_VERSION
+#[derive(Debug, PartialOrd, Eq, PartialEq, Ord)]
+pub struct SdkVersion {
+    major: usize,
+    minor: usize,
+    patch: usize,
+    release_type: ReleaseType,
+}
+
+/// Represents the release type which might be present in a version string,
+/// for example: 9.0.0-alpha.0.
+/// Release types are also comparable to each other, using the rules:
+/// alpha < beta < release.
+#[derive(Debug, Eq, PartialEq, Ord, PartialOrd)]
+enum ReleaseType {
+    Alpha(usize),
+    Beta(usize),
+    Release,
+}
+
+impl SdkVersion {
+    pub const fn new(major: usize, minor: usize, patch: usize) -> Self {
+        Self {
+            major,
+            minor,
+            patch,
+            release_type: ReleaseType::Release,
+        }
+    }
+
+    /// Attempts to parse a version string and returning a [`ParseIntError`] if it contains any
+    /// non-numerical characters apart from dots.
+    /// If a version part is not provided, 0 will be assumed.
+    /// For example:
+    /// 1.2 -> 1.2.0
+    /// 1   -> 1.0.0
+    pub fn try_parse(input: &str) -> Result<Self, ParseIntError> {
+        let mut split = input.split(".");
+        let major = split
+            .next()
+            .map(str::parse)
+            .transpose()?
+            .unwrap_or(DEFAULT_VERSION_IF_EMPTY);
```

I think if we encounter a version we don't understand we should just return an error instead of a default version, the caller can then still decide to map it to a default.

---

The idea was to bubble up errors if we don't understand the version using `transpose()?` but assume a default value for comparisons if there is just a part missing. e.g. If someone only specifies "2", we can treat is as "2.0.0" for comparisons.

Maybe it is more clear to have this:
```
split
            .next()
            .unwrap_or("0")
            .map(str::parse)
```

---

Okay makes sense, I was thinking that an invalid e.g. empty string now produces a version number, instead of an error, but I guess that is fine.